### PR TITLE
Checkbox for multiple fields selection

### LINKED ISSUE #2273: Import CSV: Checkbox for multiple fields selection
**Problem to solve**
When a user uploads a CSV file, Timesketch checks if the mandatory CSV headers are present or not. With the PR #2261 we partially solved this problem, i.e., the user can now map the missing headers with some existing ones. However, descriptive fields, such as "message" header, should be mapped to multiple fields to give the analyst the possibility to grab more information directly from that field.

**Possible solution**
The workflow is the same as #2261. However, a missing field can be mapped in two different manners: (1) using a drop down menu, or (2) using a list of check boxes. The image below shows a possible interface;

image

When the user clicks on the button 'submit', the server will receive the following parameters:
* File's metadata,
* Headers mapping, and
* Delimiter.
The headers mapping is a list of dictionaries with the following structure:
* target : the name of the header that we want to insert into the CSV,
* source : the list of headers that we want to rename or combine. It may be null if we want to insert a new column by renaming/combining any columns,
* default_value : in case the previous value is null, then this expresses the default value of the new column.

For example, let's analyze the headers_mapping like the one below:

```
headers_mapping = [
                   {"source" : "datetime", "target" : ["data_UtcTime"], "default_value" : None},
                   {"source" : "message", "message" : ["event_ID", "event_ID_desc"], "default_value" : None},
                   {"source" : "timestamp_desc", "target" : None, "default_value" : Creation_time},
                   ] 
```
It means:

1. Rename column `data_UtcTime` into `datetime`.
2. Create column `message` combining `["event_ID", "event_ID_desc"]` columns' values.
3. Create column `timestamp_desc`. Fill this column's values with a default value equal to `Creation_time`.

We decided to:
* Rename a column if the source value in the python dictionary has only one element (first dictionary of the list),
* Create a column if the source value of a dictionary consists of multiple headers or it is none.

**Additional context**
This Issue extends #2261.

### PR DESCRIPTION
**Pull request by marcobrotto**
# Import CSV: Checkbox for multiple headers selection and CSV preview
`closes #2273`

This PR closes the Issue #2273  and extends the PR #2261.

As discussed in the previous PR, 

## Check-boxes for multiple headers selection
This section explains how a user can map a missing mandatory header. In a few words, there are two different manners: using a drop down menu, or using a list of checkboxes.

MappingFinished

* When choosing the mapping from the dropdown menu, the user has two options: 
    * Map the missing header with an existing field found in the CSV, or 
    * Instruct the server to create a new column with a default value specified by the user.
For example, in the above figure, the user mapped the `datetime` header to the `data_UtcTime` one. For the timestamp_desc field he decided to create a new column. The Server will interpret these mapping in this way: it will rename the CSV header data_UtcTime into datetime, and it will create the timestamp_desc column with a default value “sysmon_event”.
* When choosing the mapping from the checkboxes list, the user can select multiple CSV headers that the server will combine into a single column named as the mandatory header. The server will not delete the selected columns. For example, in the above figure, the user mapped the field `message` to 4 different headers. The server will create a new column named `message` where each row will contain the combination of the values of the selected fields.  

## Preview of the CSV file

This PR introduces 2 buttons that (1) show a guide that helps the user to map the mandatory missing headers, and (2) have a preview of the CSV file.The preview of the CSV file is a table that shows the current status of the mapping, and how the imported file will look like after that Timesketch will elaborate it. In particular, this feature is important because it allows a user to:
* Understand what the *current* mapping looks like.
* Understand the content of each column. In fact, the user can add or remove (only from the preview) whatever column found in the CSV.

The colors of the headers in the preview show when a field is already present in the CSV (blue),  or when its mapping is incomplete (red) or done (green).

The image below shows an example of a preview of a CSV file.

image

## Front-end: major modifications:

The file that I modified in the frontend is UploadedForm.vue. The following sections explains (i) the properties that we added (included the computed properties), the new methods and the major new blocks of the HTML code.

### New properties
The new properties of the Vue object are:
* `valuesString`: raw string that we read from the uploaded CSV. This variable is mainly used to extract the values of each single header and display it in the CSV preview.
* `showHelperFlag`, `showPreviewFlag` and `showAddColumnFlag`: they are used as flags to display the preview of the CSV or the helper menu.
* `checkedHeaders`: list of the headers the user can display in the CSV preview.
* `staticNumberRows`: maximum number of rows that we want to display in CSV preview.
* `colors`: CSS property of the table, i.e., the header's style of the CSV preview.

We modified the property `mandatoryHeaders`: it is an array of dictionaries with these keys: the name of the mandatory header, and the type of mapping we want to use for it. For example, an entry of this list could be `{name : "timestamp_desc", type : "single"}`, meaning that we want to map the timestamp_desc header with only one field found in the CSV.

This structure makes the code more flexible. In fact, if we want to add a new mandatory header we just need to push a new value inside this list, as well as if we want to modify the type of a mandatory header, then we just change the respective entry of the array.

### New computed properties

* `numberRows`: the number of rows that we display in the CSV preview. Since the number of rows read by the parser could be less than `staticNumberRows` (see above), we need this parameter to avoid reading empty lines (e.g., number of lines of the file is 10 and `staticNumberRows` is 15).
*  `allHeaders`: list of all headers found in the CSV AND list of mandatory headers (no duplicates)
* `headersTable`: data structure that stores the table that we display as the preview of the CSV. Probably the most important data structure of this PR.

### New methods

* `showPreview`, `showAddColumn` and `showHelper`: functions to change the value of the flags to display the helper or the CSV preview.

The main modifications on this section involve the method `changeHeaderMapping`: in particular, the source header is now an array (not anymore a string) that stores the list of the CSV headers that the server will map to the mandatory one.

### Overview on the structure of the HTML

```
<form>
    <div v-if="extension === 'csv'">

        <!-- List of button: showHelper, showPreview, addColumnsToPreview -->
        <input type="button" @click="showHelper">...

        <!-- Dynamically generation of the preview of the CSV file -->
        <div v-if="showPreviewFlag">...</div>

        <!-- Helper menu -->
        <div v-if="showHelperFlag">...</div>

        <!-- Headers mapping -->
        <div v-for="header in missingHeaders" :key="header.name">
            <!-- Dropdown menu: 1:1 mapping OR create new header -->
            <div v-if="header.type==='single'">...</div>

            <!-- List of checkboxes: 1:N mapping -->
            <div v-if="header.type==='multiple'">...</div>
        </div>

        <!-- CSV delimiter selection -->
        <div v-for="(v, key) in delimitersList" :key="key">...</div>

    </div>

    <div v-if="errors.length === 0">
    <!-- submit form -->
    ...
    </div>
    <div v-else>
        <!-- show errors -->
        ...
    </div>
</form>

```

## Back-end: major modifications:
We slightly modified the back-end to manage the new structure of headers_mapping. We mainly modified the functions in `utils.py` according to the following algorithm:

```
def rename_headers(df, headers_mapping):
    for mapping in headers_mapping:
        if not mapping["source"]:
            # create new column with a given default value
            df.create_column(name=mapping["target"], def_value=mapping["default_value"])
        elif len(mapping["source"]) > 1:
            # concatanete multiple source headers into a new one
            concat_values = ...
            df.create_column(name=mapping["target"], values=concat_values)
        else:
            # rename the header
            df.rename(old_header = mapping["source"][0], new_header = mapping["target"])
    return df
```

## Checks

- [x] All tests succeed.
- [x] 2 Unit tests added and 2 removed (all passed)

### CODE REVIEW

File: timesketch/frontend/src/components/Common/UploadForm.vue [hunk 1]
```diff
@@ -146,60 +248,220 @@ export default {
...
+        }
+        else{
+          listSelectedHeaders.push(source)
+        }
+        this.headersMapping = this.headersMapping.filter(mapping => mapping["target"] !== target)
+        if (listSelectedHeaders.length > 0)
+          this.headersMapping.push({
+            target : target,
+            source : listSelectedHeaders,
+            default_value : defaultValue // leave snake case for python server code
```
Comment (berggren) [inline timesketch/frontend/src/components/Common/UploadForm.vue:458]:
```suggestion
            default_value : defaultValue
```

**Review by berggren — CHANGES_REQUESTED** [review]
Logic looks good. I added a few comments around style but nothing major.
Can you enable the "Vetur" extension in VSCode and have it auto format your PR? The coding style is different from the rest of the code base and it would be nice to have it consistent.

File: timesketch/lib/utils.py [hunk 2]
```diff
@@ -183,16 +183,25 @@ def check_mapping_errors(headers, headers_mapping):
...
-                    mapping["source"], ", ".join(headers))
-            )
-            # update the headers list that we will substitute
-            candidate_headers.append(mapping["source"])
+            # 3. Check if any of the headers specified in headers mapping
+            # is in the headers list
+            for source in mapping["source"]:
+                if source not in headers:
+                    raise RuntimeError(
+                    "Value specified in the headers mapping not found in the CSV\n"
```
[the remark this answers is not in the document]
Comment (marcobrotto) [reply timesketch/lib/utils.py:191]: Got it! Thank you for the feedback!

**Review by berggren — APPROVED** [review]
LGTM

**Pull request merged**